### ISSUE TITLE
Annotation for resetting context, threads and containers for integration tests

### ISSUE THREAD

**Issue by R4V3NN0** [quarkusio/quarkus#26818]
### Description

Before I switched to Quarkus, I used Spring and Spring has a feature which I really miss in Quarkus. Dont get me wrong I love working with Quarkus.

But Spring has the @DirtiesContext annotation and its very valuable in integration tests. I know that Quarkus has @TestTransaction but this seems not to work all the time. It worked fine until we integrated our database which also uses transactions and at this point @TestTransaction stopped working. In Spring this was never a problem but in Quarkus we have now the problem that our integration tests somehow depend on each other because @TestTransaction does not reset the context correctly. 

Hopefully I explained the problem well, so maybe we could implement @DirtiesContext for Quarkus which restarts the context and restarts all dev-services.

### Implementation ideas

Ideally we would implement a @DirtiesContext annotation which restarts the context, kills all active threads - so every thread which was started gets killed and does not effect the next test - and restarts all containers which were started by dev-services or Testcontainers.

**Comment by geoand** [comment]
Which kinds of Quarkus test are you talking about specifically?

**Comment by R4V3NN0** [comment]
Im talking about integration tests, so tests which are annotated with @QuarkusTest and where the whole context is spinned up. Sometimes we even spin up external services for more complete system tests.

**Comment by geoand** [comment]
You can already do this manually by using a test profile: https://quarkus.io/guides/getting-started-testing#testing_different_profiles

**Comment by R4V3NN0** [comment]
As far as I have understood this we need to write a profile for each @Quarkustest to restart the context. Does this mean that we have to write 1000 profiles when we have 1000 @QuarkusTest? Or am I getting this wrong?

**Comment by geoand** [comment]
Indeed you would have to

**Comment by R4V3NN0** [comment]
I think it would be nice if we could implement an addition to the @QuarkusTest annotation like @QuarkusTest(ResetStrategy.AfterEach) so the context would be restartet after each test. Then you could have AfterEach, AfterClass or None something like in JUnit. 
The way with @TestTransaction is nice but it is not really an option in bigger projects. We have like 400 @QuarkusTest and thats to much for adding a Testprofile for each.

**Comment by geoand** [comment]
Thanks for the feedback. We'll definitely consider it

**Comment by mihaipoenaru** [comment]
I would like something like this too. Something like Spring's annotation `DirtiesContext` that requires a test class to start with a "clean context".

**Comment by geoand** [comment]
There is no such thing a context restart in Quarkus. If we do something, it will be the exact same thing that we do when switching test profiles

**Comment by mihaipoenaru** [comment]
Ah, fair enough. In the meantime, I've already solved my issue. It was the bug that was just solved with `2.10.3`  for #26594 . I thought I messed up my tests and I had to "restart the context" in order to clear the mocks properly. Shows how much I know. 🤡

**Comment by geoand** [comment]
Yeah, that was an oversight on our part

**Comment by famod** [comment]
Is this primarily about cleaning up data from a previous test?

If you have many tests, fully restarting everything after each test will be no fun in terms of runtime!

**Comment by geoand** [comment]
> If you have many tests, fully restarting everything after each test will be no fun in terms of runtime!

Yeah, it will be very very slow...

**Comment by R4V3NN0** [comment]
Yes its primarly about cleaning data and it would be very slow if we would do that for every single test but it would still be nice if it could work like in spring. Even if it could be very slow if you use it at all tests. I think it is cleaner and safer than trying to clear old data by myself especially at massive multithreading with multiple async processes.

**Comment by deathwaiting** [comment]
This is a bit late reply, 
But if it is just reseting test data, spring actually provides the `@Sql` annotation, which runs a sql script before/after test classes or methods.
We usually depend on this in our spring projects , and no context re-run is needed 

I actually made a similar annotation in my quarkus project using CDI bean injection, so, it will be nice addition for integration tests.

**Comment by geoand** [comment]
> I actually made a similar annotation in my quarkus project using CDI bean injection, so, it will be nice addition for integration tests

Can you show what that looks like @deathwaiting?

**Comment by deathwaiting** [comment]
> > I actually made a similar annotation in my quarkus project using CDI bean injection, so, it will be nice addition for integration tests
> 
> Can you show what that looks like @deathwaiting?

@geoand 

`
    @Test
    @Sql(executionPhase = BEFORE_TEST_METHOD, scripts ="sql/test_data.sql")
    @Sql(executionPhase = AFTER_TEST_METHOD, scripts ="sql/clear_database.sql")
    public void getEventsList(){
        //test logic
    }`
    
    springs annotation targets both classes and methods, in the case of annotating a test class, it will run for each test method inside it

**Comment by geoand** [comment]
Great, thanks

**Comment by HerrDerb** [comment]
Is there already something similar? I have the feeling that the my beans in my container are getting messed up with @InjectMock and @InjectSpy

**Comment by geoand** [comment]
That should not be the case. If they are, we'll need a sample application that shows this problem so we can address it

**Comment by GustavoAngeloDS** [comment]
Hi guys, do you plan to work on this? 
Are we still not having a viable alternative for this scenario?

**Comment by geoand** [comment]
It's in the backlog, but it's not high priority

**Comment by mkouba** [comment]
I think that it would actually make sense to introduce an annotation that would imply that a test needs to "reset" the contexts (no matter what it means...) before it's executed. `@DirtiesContext` is IMO not a good name, but maybe something like `@ResetContexts`?

Implementation-wise - all extensions could contribute to the behavior - I can image ArC could destroy application and singleton contexts (that's a one-liner), `quarkus-datasource` can reset the database (this feature is already implemented for Dev UI), etc.

It might be slow but definitely faster than `@TestProfile` which requires a full app restart.

`@DirtiesContext` has also options to define _when_ the context is reset, we might need this as well...

**Comment by geoand** [comment]
What use case are you thinking about @mkouba ?

**Comment by mkouba** [comment]
> What use case are you thinking about [@mkouba](https://github.com/mkouba) ?

Typically, when you need to test a business logic that involves multiple transactions (where `@TestTransaction` does not really help), or when you need to test various states of an `@ApplicationScoped`/`@Singleton` component that does not expose a "clear/reset" method (caches, registries, etc.).

**Comment by mkouba** [comment]
Hm, maybe an annotation is not the best option here. Maybe we could introduce an interface like:

```java
interface TestState {

   void reset();

}
```

And let the specific extensions to implement this interface.

A user could then do something like:


```java
class MyTest {
   
   @Inject @All
   List<TestState> states; // used to reset all possible states

   @Inject
   DatasourceTestState db; // provided by quarkus-datasource, used to reset all databases
   
   @Inject
   ArcTestState cdi; // provided by quarkus-arc, used to reset application/singleton context
    
    @BeforeEach // or @AfterEach or even inside the test method
    void init() {
       states.forEach(TestState::reset);
       // or even
       db.reset();
       cdi.reset();
    }
}
```
@geoand WDYM?

**Comment by geoand** [comment]
I dunno, it's hard to know without concrete use cases.

**Comment by mkouba** [comment]
> I dunno, it's hard to know without concrete use cases.

Fair enough.

@R4V3NN0 @HerrDerb What do you think about the proposal [mentioned above](https://github.com/quarkusio/quarkus/issues/26818#issuecomment-3253436167)?

**Comment by holly-cummins** [comment]
There's definitely an intersection with some of the work we want to do around making reuse semantics for dev services clearer. 
(For dev services, the semantics vary between different services, making it hard for users to know what to expect. For the implementer of a service it can be hard to actually implement the intended semantics without bugs.) Having dev services able to share the same 'knob' as other aspects of the state would have some advantages. 

I guess, looking at @mkouba's proposal above, there's a tension between speed-of-execution and usability (or concision, more precisely).  The proposal seems a bit verbose and a like it might be non-obvious for users to figure out how to do it, but the alternative of a blanket `ResetStateBeforeAll` annotation would be slower, and some users might want finer control.

**Comment by mkouba** [comment]
Yes, so if you don't care about the details then you can just do:
```java
class MyTest {
   
   @All
   List<TestState> states;

   @Test
   void mySuperTestMethod() {
      states.forEach(TestState::reset);
      ...
   }
}
```
I can even imagine an "aggregate" interface like `TestStates`:
```java
class MyTest {
   
   @Inject
   TestStates states;

   @Test
   void mySuperTestMethod() {
      states.reset();
      ...
   }
}
```
And yes, you can use the specific impls to speedup the reset and finer control. An extra bonus is that you can document the impls to clarify what exactly is being reset.

I didn't really think about dev services because from user POV it might make more sense to think about higher abstractions. For example, if you use `quarkus-datasource` in your app, you can reset the DB now in the Dev UI (and under the hood the Flyway, Liquibase or Hibernate APIs are used). So it's super easy to implement the `DatasourceTestState` that would reuse existing functionality and the user can just `@Inject DatasourceTestState db` in the test and call `db.reset()` if needed.

I can also imagine that we could add `@ResetTestStates` that would just execute the `states.forEach(TestState::reset)` in the `@BeforeEach` callback 🤷 .

**Comment by HerrDerb** [comment]
@mkouba that has been a while 🙂 thanks for picking it up though.
From my side there is no need anymore. As I needed a clean reset also for container tests, I started to work with a rest endpoint for testing only (protected with a annotation that disables the endpoint in prod, which actually might be a valuable addition to quarkus 🤔) to reset all states (truncate db table, delete bean states etc.). Having this endpoint, nothing prevents me to also use it for regular service tests.

**Comment by mkouba** [comment]
So I created a [simple POC](https://github.com/mkouba/quarkus/tree/test-state) and it turns out that _we cannot destroy the CDI application/singleton contexts_ since there are extensions that might get broken because they don't expect this to happen (for example hibernate-orm). 

Developers can still destroy specific beans if needed (`jakarta.enterprise.inject.Instance.destroy(T)` for any normal scoped bean and `Arc.container().getActiveContext(Singleton.class).destroy(myBean)` for `@jakarta.inject.Singleton`).

In this POC there are both the `TestState` API (implemented by extensions and consumed by developers) and the `ResetTestState` annotation (with default behavior to reset all during `@BeforeEach`). The `TestState`  is implemented for `quarkus-datasource` and `quarkus-cache`. And it seems to work well. I find esp. the datasource impl (which works with Hibernate, Flyway and Liquibase) quite useful.

I can turn this POC into an experimental feature quite easily.
